### PR TITLE
Adding tSZ

### PR DESCRIPTION
**Pull request by wcoulton**
I've added pressure profiles from 1109.3711.

### CODE REVIEW

File: hmvec/cosmology.py [hunk 1]
```diff
@@ -4,6 +4,9 @@
 import camb
 from camb import model
 import scipy.interpolate as si
+import scipy.constants as constants
+from scipy.integrate import simps
+import tqdm
```
Comment (msyriac) [inline hmvec/cosmology.py]: Is this used anywhere?

File: hmvec/fft.py [hunk 2]
```diff
@@ -52,7 +52,7 @@ def fft_integral(x,y,axis=-1):
 def analytic_fft_integral(ks): return np.sqrt(np.pi/2.)*np.exp(-ks**2./2.)*ks
 
 
-def generic_profile_fft(rhofunc_x,cmaxs,rss,zs,ks,xmax,nxs):
+def generic_profile_fft(rhofunc_x,cmaxs,rss,zs,ks,xmax,nxs,doMassNorm=True):
```
Comment (msyriac) [inline hmvec/fft.py]: Super minor comment -- don't use CamelCase for Python variables. do_mass_norm will be more consistent with the rest of this repo and also with the majority of Python codes that try to adhere to style guidelines.

File: hmvec/hmvec.py [hunk 3]
```diff
@@ -99,19 +100,36 @@ def __init__(self,zs,ks,ms=None,params={},mass_function="sheth-torman",
...
         self.uk_profiles = {}
+        self.pk_profiles = {}
         if not(skip_nfw): self.add_nfw_profile("nfw",numeric=nfw_numeric)
 
     def _init_cosmology(self,params,halofit):
         Cosmology._init_cosmology(self,params,halofit)
         self.Pzk = self._get_matter_power(self.zs,self.ks,nonlinear=False)
         if halofit is not None: self.nPzk = self._get_matter_power(self.zs,self.ks,nonlinear=True)
 
+    def update_param(self,param,value,halofit=None):
```
Comment (msyriac) [inline hmvec/hmvec.py]: This function does not exist on master anymore and should be removed.

File: hmvec/hmvec.py [hunk 4]
```diff
@@ -255,8 +273,76 @@ def add_battaglia_profile(self,name,family=None,param_override={},
...
+        cgs = rvirs/rgs
+
+        sigmaT=constants.physical_constants['Thomson cross section'][0] # units m^2
+        mElect=constants.physical_constants['electron mass'][0] / default_params['mSun']# units kg
+        # c^2 and sigma cancel units so dont rescale.
+        # There is a G in pressure, units m3/kg/s2. But converted to Mpc, mSun, s
+        # P_200 = G M rho_crit 200 /(2* r_200)
+
+        ks,pkouts = generic_profile_fft(presFunc,cgs,rgs[...,None],self.zs,self.ks,xmax,nxs,doMassNorm=False)
+        self.pk_profiles[name] = pkouts.copy()*4*np.pi*(sigmaT/(mElect*constants.c**2))*(r200critz**3*((1+self.zs)**2/self.h_of_z(self.zs))[...,None])[...,None]# *(r200critz**3*((1+self.zs)**2/self.comoving_radial_distance(self.zs)**1)[...,None])[...,None] # *(r200critz**3*((1+self.zs)**2/self.comoving_radial_distance(self.zs)**3)[...,None])[...,None]
```
Comment (msyriac) [inline hmvec/hmvec.py]: Has the unit mismatch been fixed? If yes, please remove the comment.

File: hmvec/hmvec.py [hunk 5]
```diff
@@ -387,6 +470,7 @@ def add_hod(self,name,mthresh=None,ngal=None,corr="max",
         self.hods[name]['bg'] = self.get_bg(Ncs,Nss,self.hods[name]['ngal'])
         self.hods[name]['satellite_profile'] = satellite_profile_name
         self.hods[name]['central_profile'] = central_profile_name
+        self.hods[name]['mthres'] = np.log10(mthresh[:,None])
```
Comment (msyriac) [inline hmvec/hmvec.py]: Please name it log10mthresh to avoid confusion in code.

File: hmvec/hmvec.py [hunk 6]
```diff
@@ -429,15 +519,36 @@ def get_power_1halo(self,name="nfw",name2=None):
...
         mnames = self.uk_profiles.keys()
         hnames = self.hods.keys()
-        
-        if (name in mnames) and (name2 in mnames): square_term = self._get_matter(name)*self._get_matter(name2)
-        elif (name in hnames) and (name2 in hnames): square_term = self._get_hod_square(name)
-        elif (name in hnames) and (name2 in mnames): square_term = self._get_hod(name)*self._get_matter(name2)
-        elif (name2 in hnames) and (name in mnames): square_term = self._get_hod(name2)*self._get_matter(name)
-        else: raise ValueError
+        pnames =self.pk_profiles.keys()
+        dampingTerm=1.
```
Comment (msyriac) [inline hmvec/hmvec.py]: Please fix camelCase

File: hmvec/hmvec.py [hunk 7]
```diff
@@ -429,15 +519,36 @@ def get_power_1halo(self,name="nfw",name2=None):
...
+        if (name in hnames) and (name2 in hnames): 
+            square_term = self._get_hod_square(name)
+            dampingTerm *= np.exp(-(self.ks/self.p['kstar_damping'])**2.)
```
Comment (msyriac) [inline hmvec/hmvec.py]: Why are there multiple factors of the damping term? Why not one overall factor for the whole 1-halo term?

File: hmvec/hmvec.py [hunk 8]
```diff
@@ -429,15 +519,36 @@ def get_power_1halo(self,name="nfw",name2=None):
...
+                    dampingTerm *= np.exp(-(self.ks/self.p['kstar_damping']))
+                elif nm in mnames:
+                    square_term *= self._get_matter(nm)
+                    dampingTerm *= np.exp(-(self.ks/self.p['kstar_damping']))
+                elif nm in pnames:
+                    square_term *= self._get_pressure(nm)
+                else: raise ValueError
+
+
+        # if (name in mnames) and (name2 in mnames): square_term = self._get_matter(name)*self._get_matter(name2)
```
Comment (msyriac) [inline hmvec/hmvec.py]: Please remove comments.

File: hmvec/tinker.py [hunk 9]
```diff
@@ -58,7 +59,7 @@ def f_nu(nu,zs,delta=200.,norm_consistency=True,
     gamma = gamma0 * (1+zs)**(-0.01)
     unnormalized = (1. + (beta*nu)**(-2.*phi))*(nu**(2*eta))*np.exp(-gamma*nu**2./2.)
     if norm_consistency:
-        izs,ialphas = np.loadtxt("alpha_consistency.txt",unpack=True) # FIXME: hardcoded
+        izs,ialphas = np.loadtxt("/Users/wcoulton/Downloads/Software/hmvec/data/alpha_consistency.txt",unpack=True) # FIXME: hardcoded
```
Comment (msyriac) [inline hmvec/tinker.py]: Please revert this. Also, we should be testing with Sheth-Torman (which gives the best fit to halofit), so tinker is not really used anyway.

File: hmvec/hmvec.py [hunk 10]
```diff
@@ -452,6 +563,11 @@ def _get_term(iname):
                 rterm1 = self._get_matter(iname)
                 rterm01 = self._get_matter(iname,lowklim=True)
                 b = 1
+            elif iname in self.pk_profiles.keys():
+                rterm1 = self._get_pressure(iname)
+                rterm01 = self._get_pressure(iname,lowklim=True)
+                print ('Check the consistency relation for tSZ')
+                b = rterm01 =0
```
Comment (msyriac) [inline hmvec/hmvec.py:534]: I don't understand why this should be 0. What should the low-k limit of the yy power spectrum be?

**Review by msyriac — CHANGES_REQUESTED** [review]
Mostly looks good but some changes needed. Please see comments.

File: hmvec/cosmology.py (hunk 1 above)
Comment (msyriac) [reply hmvec/cosmology.py]: The question is about both tqdm and simps.

Comment (wcoulton) [reply hmvec/cosmology.py]: Not anymore. I've removed them.

File: hmvec/fft.py (hunk 2 above)
Comment (wcoulton) [reply hmvec/fft.py]: Great thanks for letting me know. I've renamed the variables.

File: hmvec/hmvec.py (hunk 3 above)
Comment (wcoulton) [reply hmvec/hmvec.py]: I've removed it.

File: hmvec/hmvec.py (hunk 4 above)
Comment (wcoulton) [reply hmvec/hmvec.py]: Yes. The comments have been removed

File: hmvec/hmvec.py (hunk 5 above)
Comment (wcoulton) [reply hmvec/hmvec.py]: Done.

File: hmvec/hmvec.py (hunk 6 above)
Comment (wcoulton) [reply hmvec/hmvec.py]: Done.

File: hmvec/hmvec.py (hunk 7 above)
Comment (wcoulton) [reply hmvec/hmvec.py]: After our discussion, I've altered this so all of the terms have the same damping factor (as you had before).

File: hmvec/hmvec.py (hunk 8 above)
Comment (wcoulton) [reply hmvec/hmvec.py]: Done.

File: hmvec/tinker.py (hunk 9 above)
Comment (wcoulton) [reply hmvec/tinker.py]: Done.

**Review by msyriac — APPROVED** [review]

**Pull request merged**